presto-main/src/main/java/com/facebook/presto/operator/VariableWidthBlockEncodingBuffers.java

```diff
@@ -0,0 +1,257 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator;
+
+import com.facebook.presto.spi.block.AbstractVariableWidthBlock;
+import com.google.common.annotations.VisibleForTesting;
+import io.airlift.slice.SliceOutput;
+import org.openjdk.jol.info.ClassLayout;
+
+import java.util.Arrays;
+
+import static com.facebook.presto.operator.UncheckedByteArrays.setIntUnchecked;
+import static com.facebook.presto.spi.MoreByteArrays.setBytes;
+import static io.airlift.slice.SizeOf.SIZE_OF_INT;
+import static io.airlift.slice.SizeOf.sizeOf;
+import static java.lang.Math.max;
+import static sun.misc.Unsafe.ARRAY_INT_INDEX_SCALE;
+
+public class VariableWidthBlockEncodingBuffers
+        extends BlockEncodingBuffers
+{
+    @VisibleForTesting
+    public static final int POSITION_SIZE = Integer.BYTES + Byte.BYTES;
+
+    private static final String NAME = "VARIABLE_WIDTH";
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(VariableWidthBlockEncodingBuffers.class).instanceSize();
+    private static final int ESTIMATED_ELEMENT_SIZE = 8;
+
+    // The buffer for the slice for all incoming blocks so far
+    private byte[] sliceBuffer;
+
+    // The address that the next slice will be written to.
+    private int sliceBufferIndex;
+
+    // The buffer for the offsets for all incoming blocks so far
+    private byte[] offsetsBuffer;
+
+    // The address that the next offset value will be written to.
+    private int offsetsBufferIndex;
+
+    // The last offset in the offsets buffer
+    private int lastOffset;
+
+    public VariableWidthBlockEncodingBuffers(int initialPositionCount)
+    {
+        super(initialPositionCount);
+    }
+
+    @Override
+    public void resetBuffers()
+    {
+        bufferedPositionCount = 0;
+        sliceBufferIndex = 0;
+        offsetsBufferIndex = 0;
+        lastOffset = 0;
+        resetNullsBuffer();
+    }
+
+    @Override
+    public void accumulateRowSizes(int[] rowSizes)
+    {
+        int[] positions = getPositions();
+        for (int i = 0; i < positionCount; i++) {
+            rowSizes[i] += decodedBlock.getSliceLength(positions[i]);
+            rowSizes[i] += POSITION_SIZE;
+        }
+    }
+
+    @Override
+    protected void accumulateRowSizes(int[] offsetsOffsets, int positionCount, int[] rowSizes)
```

Comment:
offsetsOffsets name doesn't read well; why not just `offsets`?

Reply:
It will be confused with offsets array in other BlockEncodingBuffers classes because they all have this method. I currently renamed all of them back to positionOffsets. But I think topLevelOffsets may make more sense?

Reply:
> Some initial comments about `OptimizedPartitionedOutputOperator`.
> 
> Here is my high-level understanding, feel free to correct me if I misunderstand anything:
> 
> * The current implementation  of `PartitionedOutputOperator` implements an inner class `PagePartitioner`.
>   It maintains an output `PageBuilder` for each output partition. For each input row, calculate the output partition of it, and `appendRow` to the output `PageBuilder` and later serialized to bytes. Thus, data has to be first append into `PageBuilder` and later serialized, which is inefficient.
> * The new `OptimizedPartitionedOutputOperator` tries to do "partition-and-serialize". As a result, its corresponding `PagePartitioner` maintains an output `BlockEncodingBuffer` for each output partition (stored in `PartitionData`).
>   To do "partition-and-serialize", it did in a two-phase fashion:
>   
>   * (Output Register Phase) In the first pass, go through the rows in the input page, calculate the target partition and "register" it (calling `PartitionData.appendPosition`)
>   * (Serialization Phase) In the second pass, deserialize the data at marked positions (calling `PartitionData.appendData`)
>     I assume this the motivation for this two-phase implementation is because the serialization phase can be done in a tight loop (Aria philosophy: don't distract in tight loop 😃 )

Correct. We want to do the data appending column by column instead of row by row. Embedding the destination partition calculation in the row by row fashion is even worse. In the "second pass" you mentioned, we also calculate the row sizes for the whole block before appending data to avoid size checks per position. Result of these trick is optimized tight loops.

> Some questions:
> 
> * Why the name is `BlockEncodingBuffer`?  (instead of, for example, `SerializedBlockBuffer`, etc).

Regarding the name, I just copy from BlockEncoding because they do similar things. I'm open to renaming suggestions.

> * I would love to understand a bit more about `DecodedBlockNode`. My high-level understanding is to achieve high performance when navigating through "delegate blocks" such as RLE/Dictionary Block. What's the relationship with flattened block and this "decoded block"?

I added some explanation to the code. The purpose of it is to avoid decoding each incoming block multiple times when the block is large and has to be appended in multiple batches. Block decoding is expensive.